SU2_CFD/src/solvers/CSpeciesFlameletSolver.cpp

```diff
@@ -0,0 +1,513 @@
+
+      if (n_not_in_domain_global > 0)
+        cout << " Initial condition: Number of points outside of table domain: " << n_not_in_domain_global << " !!!"
+             << endl;
+
+      if (n_not_iterated_global > 0)
+        cout << " Initial condition: Number of points in which enthalpy could not be iterated: " << n_not_iterated_global
+             << " !!!" << endl;
+    }
+  }
+}
+
+void CSpeciesFlameletSolver::SetPreconditioner(CGeometry* geometry, CSolver** solver_container, CConfig* config) {
+  const bool variable_density = (config->GetKind_DensityModel() == INC_DENSITYMODEL::VARIABLE);
+  const bool implicit = (config->GetKind_TimeIntScheme_Flow() == EULER_IMPLICIT);
+
+  SU2_OMP_FOR_STAT(omp_chunk_size)
+  for (auto iPoint = 0u; iPoint < nPointDomain; iPoint++) {
+    /*--- Access the primitive variables at this node. ---*/
+
+    su2double Density = solver_container[FLOW_SOL]->GetNodes()->GetDensity(iPoint);
+    su2double BetaInc2 = solver_container[FLOW_SOL]->GetNodes()->GetBetaInc2(iPoint);
+    su2double Temperature = solver_container[FLOW_SOL]->GetNodes()->GetTemperature(iPoint);
+
+    su2double SolP = solver_container[FLOW_SOL]->LinSysSol(iPoint, prim_idx.Pressure());
+    su2double SolT = solver_container[FLOW_SOL]->LinSysSol(iPoint, prim_idx.Temperature());
+
+    /*--- We need the derivative of the equation of state to build the
+     preconditioning matrix. For now, the only option is the ideal gas
+     law, but in the future, dRhodT should be in the fluid model. ---*/
+
+    su2double dRhodT = 0.0;
+    if (variable_density) {
+      dRhodT = -Density / Temperature;
+    }
+
+    /*--- Passive scalars have no impact on the density. ---*/
+
+    su2double dRhodC = 0.0;
+
+    /*--- Modify matrix diagonal with term including volume and time step. ---*/
+
+    su2double Vol = geometry->nodes->GetVolume(iPoint);
+    su2double Delta =
+        Vol / (config->GetCFLRedCoeff_Species() * solver_container[FLOW_SOL]->GetNodes()->GetDelta_Time(iPoint));
+
+    /*--- Calculating the inverse of the preconditioning matrix
+     that multiplies the time derivative during time integration. ---*/
+
+    if (implicit) {
+      for (unsigned short iVar = 0; iVar < nVar; iVar++) {
+        su2double scalar = nodes->GetSolution(iPoint, iVar);
+
+        /*--- Compute the lag terms for the decoupled linear system from
+         the mean flow equations and add to the residual for the scalar.
+         In short, we are effectively making these terms explicit. ---*/
+
+        su2double artcompc1 = SolP * scalar / (Density * BetaInc2);
+        su2double artcompc2 = SolT * dRhodT * scalar / (Density);
+
+        LinSysRes(iPoint, iVar) += artcompc1 + artcompc2;
+
+        /*--- Add the extra Jacobian term to the scalar system. ---*/
+
+        su2double Jaccomp = scalar * dRhodC + Density;
+        su2double JacTerm = Jaccomp * Delta;
+
+        Jacobian.AddVal2Diag(iPoint, JacTerm);
```

Comment:
Ok, then this should be enough, each value is different so it cannot be done outside of the loop.
```suggestion
        Jacobian.AddVal2Diag(iPoint, iVar, JacTerm);
```